packages/remote-feature-flag-controller/src/remote-feature-flag-controller.ts

```diff
@@ -291,14 +332,43 @@ export class RemoteFeatureFlagController extends BaseController<
         continue;
       }
 
-      if (Array.isArray(processedValue) && thresholdValue) {
+      if (Array.isArray(processedValue)) {
+        // Validate array has valid threshold items before doing expensive crypto operation
+        const hasValidThresholds = processedValue.some(
+          isFeatureFlagWithScopeValue,
+        );
+
+        if (!hasValidThresholds) {
+          // Not a threshold array - preserve as-is
+          processedRemoteFeatureFlags[remoteFeatureFlagName] = processedValue;
+          continue;
+        }
+
+        // Check cache first, calculate only if needed
+        const cacheKey: `${string}:${string}` = `${metaMetricsId}:${remoteFeatureFlagName}`;
+        let thresholdValue = this.state.thresholdCache[cacheKey];
+
+        if (thresholdValue === undefined) {
+          // eslint-disable-next-line require-atomic-updates
+          thresholdValue = await calculateThresholdForFlag(
+            metaMetricsId,
+            remoteFeatureFlagName,
+          );
+
+          // Update state to persist the threshold
+          this.update((state) => {
```

Comment:
Nit: This will create one state change event per update, which could mean dozens of changes in a short time frame. It'll then change _again_ when we clear out stale flags.

To reduce the frequency of these state change events, we should be able to consolidate this into a single update (with all new thresholds added, and new ones removed).